Classes/common/CDTIncrementalStore.m

```diff
@@ -1571,75 +1370,73 @@ - (void)setMetadata:(NSDictionary *)metadata
     [super setMetadata:metadata];
 }
 
-#pragma mark - Database Delegates
-/**
- * Called when the replicator changes state.
- */
-- (void)replicatorDidChangeState:(CDTReplicator *)replicator
+#pragma mark - Replication Creators
+- (CDTReplicator *)replicatorThatPushesToURL:(NSURL *)remoteURL withError:(NSError **)error
 {
-    NSString *state;
-    switch (replicator.state) {
-        case CDTReplicatorStatePending:
-            state = @"CDTReplicatorStatePending";
-            break;
-        case CDTReplicatorStateStarted:
-            state = @"CDTReplicatorStateStarted";
-            break;
-        case CDTReplicatorStateStopped:
-            state = @"CDTReplicatorStateStopped";
-            break;
-        case CDTReplicatorStateStopping:
-            state = @"CDTReplicatorStateStopping";
-            break;
-        case CDTReplicatorStateComplete:
-            state = @"CDTReplicatorStateComplete";
-            break;
-        case CDTReplicatorStateError:
-            state = @"CDTReplicatorStateError";
-            break;
-        default:
-            state = @"unknown replicator state";
-            CDTLogWarn(CDTREPLICATION_LOG_CONTEXT, @"%@: %@: %@: state: %@", CDTISType,
-                       [self cleanURL:self.remoteURL], replicator, state);
-            break;
+    NSError *err = nil;
+    NSString *clean = [self cleanURL:remoteURL];
+
+    CDTReplicatorFactory *repFactory =
+        [[CDTReplicatorFactory alloc] initWithDatastoreManager:self.manager];
+    if (!repFactory) {
+        CDTLogError(CDTREPLICATION_LOG_CONTEXT, @"%@: %@: Could not create replication factory",
+                    CDTISType, clean);
+        return nil;
     }
 
-    CDTLogInfo(CDTREPLICATION_LOG_CONTEXT, @"%@: %@: %@: state: %@", CDTISType,
-               [self cleanURL:self.remoteURL], replicator, state);
-}
+    CDTPushReplication *pushRep =
+        [CDTPushReplication replicationWithSource:self.datastore target:remoteURL];
+    if (!pushRep) {
+        CDTLogError(CDTREPLICATION_LOG_CONTEXT, @"%@: %@: Could not create push replication object",
+                    CDTISType, clean);
+        return nil;
+    }
```

Comment:
I think it is possible that `repFactory` could be reused, I will check and I think we could just create it up front.
However, the `(push|pull)Rep` objects are actually single use, so we need to create them on demand.